nvflare/app_opt/flower/plugin.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tempfile
+from logging import ERROR
+from pathlib import Path
+from typing import List
+
+import pathspec
+from flwr.cli.config_utils import get_fab_metadata
+from flwr.cli.install import install_from_fab
+from flwr.common.logger import log
+from flwr.common.typing import UserConfig
+from flwr.superexec.executor import Executor as FlowerSuperExecExecutor
+from flwr.superexec.executor import RunTracker
+from typing_extensions import Optional, override
+
+from nvflare import FedJob
+from nvflare.app_opt.flower.controller import FlowerController
+from nvflare.app_opt.flower.executor import FlowerExecutor
+from nvflare.fuel.flare_api.flare_api import Session, new_secure_session
+
+SESSION_ARGS = {"username", "startup_kit_location", "debug", "timeout"}
+
+
+def _get_job_name(publisher: str, app_name: str, app_version: str) -> str:
+    """Generate the job name."""
+    # Replace invalid characters
+    app_name = app_name.replace(" ", "_")
+
+    # Return the job name
+    return f"{app_name}@{publisher}[{app_version}]"
+
+
+def _load_gitignore(directory: Path) -> pathspec.PathSpec:
+    """Load and parse .gitignore file, returning a pathspec."""
+    gitignore_path = directory / ".gitignore"
+    patterns = ["__pycache__/"]  # Default pattern
+    if gitignore_path.exists():
+        with open(gitignore_path, encoding="UTF-8") as file:
+            patterns.extend(file.readlines())
+    return pathspec.PathSpec.from_lines("gitwildmatch", patterns)
+
+
+def _locate_all_files(directory: Path) -> List[Path]:
+    """Locate all allowed files in the directory."""
+    # Allowed extensions
+    allowed_exts = ["py", "toml", "md"]
+
+    # Load gitignore
+    gitignore = _load_gitignore(directory)
+
+    # Walk through the directory
+    files = []
+    for entry in (_ for ext in allowed_exts for _ in directory.rglob(f"*.{ext}")):
+        if not gitignore.match_file(entry.relative_to(directory)):
+            files.append(entry)
+    return files
```

Comment:
Put all located files into a temp folder (e.g. /tmp/nakjsdbwejkbdf), and return this folder path. See comments in below.